api/src/utils/validate-diff.ts

```diff
@@ -0,0 +1,150 @@
+import Joi from 'joi';
+import { InvalidPayloadException } from '../index';
+import { DiffKind, SnapshotDiffWithHash, SnapshotWithHash } from '../types/snapshot';
+
+const deepDiffSchema = Joi.object({
+	kind: Joi.string()
+		.valid(...Object.values(DiffKind))
+		.required(),
+	path: Joi.array().items(Joi.string()),
+	lhs: Joi.object().when('kind', { is: [DiffKind.DELETE, DiffKind.EDIT], then: Joi.required() }),
+	rhs: Joi.object().when('kind', { is: [DiffKind.NEW, DiffKind.EDIT], then: Joi.required() }),
+	index: Joi.number().when('kind', { is: DiffKind.ARRAY, then: Joi.required() }),
+	item: Joi.link('/').when('kind', { is: DiffKind.ARRAY, then: Joi.required() }),
+});
+
+const applyJoiSchema = Joi.object({
+	hash: Joi.string().required(),
+	diff: Joi.object({
+		collections: Joi.array()
+			.items(
+				Joi.object({
+					collection: Joi.string().required(),
+					diff: Joi.array().items(deepDiffSchema).required(),
+				})
+			)
+			.required(),
+		fields: Joi.array()
+			.items(
+				Joi.object({
+					collection: Joi.string().required(),
+					field: Joi.string().required(),
+					diff: Joi.array().items(deepDiffSchema).required(),
+				})
+			)
+			.required(),
+		relations: Joi.array()
+			.items(
+				Joi.object({
+					collection: Joi.string().required(),
+					field: Joi.string().required(),
+					related_collection: Joi.string(),
+					diff: Joi.array().items(deepDiffSchema).required(),
+				})
+			)
+			.required(),
+	}).required(),
+});
+
+/**
+ * Validates the diff against the current schema snapshot.
+ *
+ * @returns True if the diff can be applied (valid & not empty).
+ */
+export function validateApplyDiff(applyDiff: SnapshotDiffWithHash, currentSnapshotWithHash: SnapshotWithHash) {
+	const { error } = applyJoiSchema.validate(applyDiff);
+	if (error) throw new InvalidPayloadException(error.message);
+
+	if (
+		applyDiff.diff.collections.length === 0 &&
+		applyDiff.diff.fields.length === 0 &&
+		applyDiff.diff.relations.length === 0
+	) {
+		return false;
+	}
+
+	if (applyDiff.hash !== currentSnapshotWithHash.hash) {
```

Comment:
Similar to validateSnapshot, how do we feel with doing `if (applyDiff.hash === currentSnapshotWithHash.hash) return false;` instead? (or `if (applyDiff.hash === currentSnapshotWithHash.hash) return;` depending on our discussion in https://github.com/directus/directus/pull/14713#discussion_r1090691487)